llmfoundry/models/inference_api_wrapper/openai_causal_lm.py

```diff
@@ -0,0 +1,239 @@
+            raise ValueError(
+                f'`encode` argument must be str or List[str], got: {type(x)}')
+
+    def decode(self, x: List[int]):
+        return self.tokenizer.decode(x)
+
+    @property
+    def pad_token_id(self):
+        return self.tokenizer.eot_token
+
+    @property
+    def eos_token_id(self):
+        return self.tokenizer.eot_token
+
+    def construct_logit_tensor(self, logprobs: Dict[str, float]):
+        tensor = torch.tensor([min(logprobs.values()) - 1] *
+                              (self.pad_token_id + 1))
+        for k in logprobs:
+            encoding = self.encode(k)['input_ids']
+            idx = encoding[0]
+            tensor[idx] = logprobs[k]
+        return tensor
+
+
+class OpenAIChatAPIEvalWrapper(InferenceAPIEvalWrapper):
+
+    def retokenize(self, tokens: List[int], cont_idxs: List[int]):
+        original_len = len(tokens)
+        retokenized_continuation = self.tokenizer.encode(
+            self.tokenizer.decode(tokens[cont_idxs[0]:cont_idxs[-1] +
+                                         1]).strip())['input_ids']
+        tokens = tokens[:cont_idxs[0]] + retokenized_continuation + [
+            tokens[-1]
+        ] * (len(tokens) -
+             len(tokens[:cont_idxs[0]] + retokenized_continuation))
+        if len(tokens) > original_len:
+            # this only happens if we were already at max seq len and the continuation got LARGER
+            tokens = tokens[-original_len:]
+            cont_idxs = list(
+                range(original_len - len(retokenized_continuation),
+                      original_len))
+        else:
+            cont_idxs = list(
+                range(cont_idxs[0],
+                      cont_idxs[0] + len(retokenized_continuation)))
+        return torch.tensor(tokens), torch.tensor(cont_idxs)
+
+    def rebatch(self, batch: Batch):
+        new_batch: Dict[str, Union[List[torch.Tensor], torch.Tensor]] = {
+            'input_ids': [],
+            'continuation_indices': [],
+            'labels': []
+        }
+        for tokens, cont_idxs in zip(batch['input_ids'],
+                                     batch['continuation_indices']):
+            tokens, cont_idxs = self.retokenize(tokens.tolist(),
+                                                cont_idxs.tolist())
+
+            assert isinstance(new_batch['input_ids'], list)
+            new_batch['input_ids'].append(tokens)
+            assert isinstance(new_batch['labels'], list)
+            new_batch['labels'].append(tokens)
+            assert isinstance(new_batch['continuation_indices'], list)
+            new_batch['continuation_indices'].append(cont_idxs)
+
+        new_batch.update({
+            k: torch.stack(new_batch[k])  # pyright: ignore
+            for k in ['input_ids', 'labels']
+        })
+
+        new_batch.update({k: v for k, v in batch.items() if k not in new_batch})
+
+        return new_batch
+
+    def eval_forward(self, batch: Batch, outputs: Optional[Any] = None):
```

Comment:
is this different from the base class implementation?

Reply:
Yes, it does the rebatching at the beginning and it also decodes the entire response at once rather than a single token at a time.

This is because ChatAPI behaves differently in that it always strips spaces from its outputs